xxh3.h

```diff
@@ -290,53 +305,54 @@ XXH_FORCE_INLINE U64x2 XXH_vec_revb(U64x2 val)
     return vec_perm(val, val, vByteSwap);
 }
 #  endif
+# endif /* XXH_VSX_BE */
 
-/* Power8 Crypto gives us vpermxor which is very handy for
- * PPC64EB.
- *
- * U8x16 vpermxor(U8x16 a, U8x16 b, U8x16 mask)
- * {
- *     U8x16 ret;
- *     for (int i = 0; i < 16; i++) {
- *         ret[i] = a[mask[i] & 0xF] ^ b[mask[i] >> 4];
- *     }
- *     return ret;
- * }
- *
- * Because both of the main loops load the key, swap, and xor it with input,
- * we can combine the key swap into this instruction.
+/*
+ * Performs an unaligned load and byte swaps it on big endian.
  */
-#  ifdef vec_permxor
-#    define XXH_vec_permxor vec_permxor
-#  else
-#    define XXH_vec_permxor __builtin_crypto_vpermxor
-#  endif
-#endif  /* XXH_VSX_BE */
+XXH_FORCE_INLINE U64x2 XXH_vec_loadu(const void *ptr)
+{
+    U64x2 ret;
+    memcpy(&ret, ptr, sizeof(U64x2));
+# if XXH_VSX_BE
+    ret = XXH_vec_revb(ret);
+# endif
+    return ret;
+}
+
 /*
- * Because we reinterpret the multiply, there are endian memes: vec_mulo actually becomes
- * vec_mule.
+ * vec_mulo and vec_mule are very problematic intrinsics:
  *
- * Additionally, the intrinsic wasn't added until GCC 8, despite existing for a while.
- * Clang has an easy way to control this, we can just use the builtin which doesn't swap.
- * GCC needs inline assembly. */
-#if defined(__clang__) && __has_builtin(__builtin_altivec_vmuleuw)
+ * The intrinsic weren't added until GCC 8, despite existing for a while.
+ * Besides, they are endian dependent, and their meaning swap depending on version.
+ * */
+# if defined(__clang__) && __has_builtin(__builtin_altivec_vmuleuw)
+/* Clang has a better way to control this, we can just use the builtin which doesn't swap. */
 #  define XXH_vec_mulo __builtin_altivec_vmulouw
 #  define XXH_vec_mule __builtin_altivec_vmuleuw
-#else
+# elif defined(__s390x__) && defined(__GNUC__) && (__GNUC__ >= 8)
+/* s390x is always big endian, there is no endian issue. */
+#  define XXH_vec_mulo vec_mulo
+#  define XXH_vec_mule vec_mule
+# else
```

Comment:
Considering [this test result](https://travis-ci.org/Cyan4973/xxHash/jobs/649629973#L210), I presume a mistake here.

cc @easyaspi314 .

Reply:
`vec_mulo` and `vec_mule` seem to be available with gcc-5 and later with `-march=z13 -mzvector`.

Reply:
Changed the policy, so that, on `s390x`, `vec_mulo` and `vec_mule` intrinsics are always used, as in the original patch. There is no `gcc` version check.

This change fixed the tests.

We may want to add more tests if we want to catch issues not detected by current test suite.